WIP: Backport FILTER implementation, from PostgreSQL 9.4 (commit b560ec1b0d).

hlinnaka opened the pull request:
We had a very simplistic implementation in parse-analysis already, which
converted the FILTER WHERE clause into a CASE-WHEN expression. That did
not work for non-strict aggregates, and didn't deparse back into a FILTER
expression nicely, to name a few problems with it. Replace it with the
PostgreSQL implementation.

TODO:

* ORCA support. It now falls back to the Postgres planner.
* I disabled the three-stage DQA plan types if there are any FILTERs

hlinnaka commented:
@vraghavan78: This will simplify things for the work to use upstream window functions implementation. Although PostgreSQL only got support for FILTER in version 9.4, if we backport that now, it would be one less difference to worry about. And it would add functionality over the current implementation, as it has some limitations. And it would fix issue https://github.com/greenplum-db/gpdb/issues/3153.

This PR is correct, as in, everything works. However, as noted in the commit message, ORCA support doesn't exist, so this causes queries with FILTERs to fall back to the Postgres planner. Also, within the Postgres planner, those three-stage aggregate plans are disabled, because they are similarly not implemented yet for FILTERs.

So I'd appreciate some help from the Optimizer folk to address those two issues. In the meanwhile, I'd like to commit this now, as it is, to get this out of the way of the merge, on the assumption that those two issues will be addressed before the release. Thoughts?

hlinnaka commented:
Ping, @vraghavan78?

hsyuan commented on src/backend/gpopt/translate/CTranslatorScalarToDXL.cpp:
```diff
@@ -1340,6 +1340,12 @@ CTranslatorScalarToDXL::PdxlnScAggrefFromAggref
+	// ORCA doesn't support the FILTER clause yet.
+	if (paggref->aggfilter)
+	{
+		GPOS_RAISE(gpdxl::ExmaDXL, gpdxl::ExmiQuery2DXLUnsupportedFeature, GPOS_WSZ_LIT("Non-Scalar Subquery"));
```
Should the error messages be "Aggregate functions with FILTER"?

pivotal-mike commented:
@vraghavan78 @oarap For visibility as this generates some work for QP

@schubert for visibility

hlinnaka commented:
Rebased, fixed the error message on ORCA fallaback, and pushed.

I think the three-stage Agg thing might get resolved by the window functions work, but would be good to make a note of that somewhere, so that we remember to revisit that before the release. And ORCA support certainly should be added at some point.
